src/Freya.Integration/Integration.fs

```diff
@@ -0,0 +1,105 @@
+[<RequireQualifiedAccess>]
+[<CompilationRepresentation (CompilationRepresentationFlags.ModuleSuffix)>]
+module OwinAppFunc =
+    
+    /// Converts an <see cref="OwinAppFunc" /> to a <see cref="Freya{T}" /> computation
+    /// to allow use of standard OWIN components within Freya.
+    /// NOTE: EXPERIMENTAL
+    [<CompiledName ("ToFreya")>]
+    let toFreya (app: OwinAppFunc) : Freya<unit> =
+        fun s -> async {
+            let! token = Async.CancellationToken
+            // Apply and mutate the OwinEnvironment asynchronously
+            do! Async.AwaitTask <| app.Invoke(s.Environment).ContinueWith<unit>((fun _ -> ()), token)
+            // Return the result as a unit value and the mutated FreyaState
+            return (), s }
+
+open Freya.Pipeline
+
+/// Type alias for the OWIN MidFunc signature.
+type OwinMidFunc =
+    Func<OwinAppFunc, OwinAppFunc>
+
+/// Provides transformation functions for converting to/from OWIN middlewares from/to Freya Pipelines.
+[<RequireQualifiedAccess>]
+[<CompilationRepresentation (CompilationRepresentationFlags.ModuleSuffix)>]
+module OwinMidFunc =
+    
+    /// Converts a Freya.Pipeline to an OWIN MidFunc.
+    [<CompiledName("FromFreya")>]
+    let fromFreya (pipeline: Freya<FreyaPipelineChoice>) =
+        OwinMidFunc(fun next ->
+            let app e =
+                // Convert to FreyaState
+                let s = { Environment = e
+                          Meta = { Memos = Map.empty } }
+                async {
+                    // Execute the pipeline
+                    let! choice, s' = pipeline s
+                    match choice with
+                    // Execute the next OwinAppFunc
+                    | Next ->
+                        return! next.Invoke(s'.Environment).ContinueWith<unit>(fun _ -> ()) |> Async.AwaitTask
+                    // Complete the Task
+                    | Halt -> return () }
+                |> Async.StartAsTask :> Task
+            OwinAppFunc app)
+
+    /// Splits a MidFunc into a before and after Freya.Pipeline.
+    [<CompiledName("SplitIntoFreya")>]
+    let splitIntoFreya (midFunc: OwinMidFunc) : Freya<FreyaPipelineChoice> * Freya<FreyaPipelineChoice> =
+        let nextWasRun = ref false
+        let resumed = new Event<unit>()
+        let nextSignal =
+            OwinAppFunc(fun _ ->
+                async {
+                    nextWasRun := true
+                    // Pause while the rest of the pipeline runs.
+                    do! resumed.Publish |> Async.AwaitEvent
+                    // Complete the Task.
+                    return ()
+                } |> Async.StartAsTask :> Task)
+        let before s =
+            async {
+                let! token = Async.CancellationToken
+                // Apply and mutate the OwinEnvironment asynchronously.
+                do! Async.AwaitTask <| midFunc.Invoke(nextSignal).Invoke(s.Environment).ContinueWith<unit>((fun _ -> ()), token)
```

Comment:
I think the problem lies here. Rather than wait for the middleware to complete, we should signal whether or not `nextWasRun`. However, I'm not sure how to detect whether or not the middleware halted. Perhaps we need another wrapper around the middleware to detect the middleware completing the `Task`?